src/gpu/generic/sycl/rnn/ref_rnn.cpp

```diff
@@ -0,0 +1,791 @@
+                break;
+            }
+        }
+        return status::unimplemented;
+    };
+
+    float gemm_iter_fwd_beta = this->is_lbr() ? 0.0f : 1.0f;
+
+    // Setup gemm PDs
+    if (aprop == prop_kind::forward) {
+        VDISPATCH_RNN_SC(create_gemm_pd(gemm_layer_fwd_pd_, n_gates * dhc,
+                                 batch, slc, {rnn_conf.states_ws_ld, 1},
+                                 {off.weights_layer[2], off.weights_layer[4]},
+                                 {rnn_conf.scratch_gates_ld, 1}, weights_type,
+                                 src_type, rnn_conf.acc_data_type, 0.0),
+                "create_gemm_pd(gemm_layer_fwd_pd_)");
+
+        VDISPATCH_RNN_SC(
+                create_gemm_pd(gemm_iter_fwd_pd_, n_gates * dhc, batch, sic,
+                        {rnn_conf.states_ws_ld, 1},
+                        {off.weights_iter[2], off.weights_iter[4]},
+                        {rnn_conf.gates_ws_ld, 1}, weights_type, src_type,
+                        rnn_conf.acc_data_type, gemm_iter_fwd_beta),
+                "create_gemm_pd(gemm_iter_fwd_pd_)");
+    }
+
+    init_scratchpad(rnn_conf.use_workspace ? 0 : workspace_size);
+    return status::success;
+}
+
+template <prop_kind_t aprop>
+status_t _ref_rnn_common_t<aprop>::init(impl::engine_t *engine) {
+    using namespace rnn_utils;
+
+    switch (pd()->cell_kind()) {
+        case dnnl_vanilla_rnn: cell_func = &class_name::cell_execution; break;
+        default: break;
+    }
+
+    grid_computation = &class_name::linear_execution;
+
+    const conf_t &rnn = pd()->rnn_conf;
+    rnn_utils::set_workspace_offsets(rnn, ws_gates_offset_, ws_states_offset_);
+
+    // IMPORTANT SYCL STUFF
+    const auto copy_kid = ::sycl::get_kernel_id<ref_rnn_copy_t>();
+    this->create_kernel(engine, copy_kid, &copy_kernel_);
+    const auto bias_kid = ::sycl::get_kernel_id<ref_rnn_bias>();
+    this->create_kernel(engine, bias_kid, &bias_kernel_);
+
+    bool gemm_ok = true;
+    auto create_nested_gemm =
+            [&](const std::shared_ptr<primitive_desc_t> &prim_desc,
+                    std::shared_ptr<impl::primitive_t> &prim) {
+                std::pair<std::shared_ptr<impl::primitive_t>, cache_state_t>
+                        pair;
+                bool gemm_ok = prim_desc->create_primitive_nested(pair, engine)
+                        == status::success;
+                prim = pair.first;
+                return gemm_ok;
+            };
+
+    gemm_ok = gemm_ok
+            && create_nested_gemm(pd()->gemm_layer_fwd_pd_, gemm_layer_fwd_);
+    gemm_ok = gemm_ok
+            && create_nested_gemm(pd()->gemm_iter_fwd_pd_, gemm_iter_fwd_);
+
+    if (!gemm_ok) return status::runtime_error;
+
+    return status::success;
+} // namespace sycl
+
+template <prop_kind_t aprop>
+gemm_sig((_ref_rnn_common_t<aprop>::gemm_primitive)) {
+    // We flip A and B here since the GEMM API is row major but the
+    // RNN code describes GEMM in column major fashion
```

Comment:
This comment doesn't seem to apply given the switch to matmul.

Reply:
Outdated comment removed.